### PR TITLE
This is the V0.1

### PR DESCRIPTION
**Pull request by dialmformart**
Need some grammar corrections, but I think that's a good first pt-br version!

### CODE REVIEW

**Comment by dialmformart** [comment]
@danielfireman Check it out!

File: performance-ptbr.md [hunk 1]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
 Avaliando a precisão do benchmark:
 
 * <http://www.brendangregg.com/blog/2018-06-30/benchmarking-checklist.html>
+
+### Melhoria do programa.
+
+A melhoria do programa costumava ser uma forma de arte, mas os compiladores ficaram melhores. Portanto, agora os compiladores podem otimizar o código direto melhor do que complicar aquele código.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
A melhoria do programa costumava ser uma forma de arte, mas os compiladores ficaram melhores. Portanto, agora os compiladores podem otimizar o código melhor do que complicar aquele código. 
```

File: performance-ptbr.md [hunk 2]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
 
 * <http://www.brendangregg.com/blog/2018-06-30/benchmarking-checklist.html>
+
+### Melhoria do programa.
+
+A melhoria do programa costumava ser uma forma de arte, mas os compiladores ficaram melhores. Portanto, agora os compiladores podem otimizar o código direto melhor do que complicar aquele código. 
+O compilador Go ainda tem um longo caminho a percorrer para corresponder ao gcc e ao clang, mas isso significa que você precisa 
+ter cuidado ao ajustar e principalmente ao atualizar as versões de Go para que seu código não se torne "pior".
+Definitivamente, há casos em que os ajustes para solucionar a falta de uma otimização específica do compilador
+façam o código se tornar mais lentos depois que ao compilador foi aprimorado.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
façam o código se tornar mais lento depois que o compilador foi aprimorado.
```

File: performance-ptbr.md [hunk 3]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+
+
+Minha implementação de cifra RC6 teve um aumento de velocidade de 10% para o loop interno apenas mudando para `encoding / binary` e` math / bits` em vez de usar uma de minhas versões feitas manualmente.
+
+Da mesma forma, o pacote `compress / bzip2` foi acelerado ao mudar para [código mais simples que o compilador conseguiu otimizar] (https://github.com/golang/go/commit/9eb219480e8de08d380ee052b7bff293856955f8)
+
+Se você estiver trabalhando em torno de um tempo de execução específico ou geração de código do compilador sempre documente sua alteração com um link para a edição anterior. este permitirá que você revisite rapidamente sua otimização depois que o bug for corrigido.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Se você estiver trabalhando em torno de um tempo de execução específico ou geração de código do compilador sempre documente sua alteração com um link para a edição anterior. Este link permitirá que você revisite rapidamente sua otimização depois que o bug for corrigido.
```

File: performance-ptbr.md [hunk 4]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+
+Lute contra a tentação de "dicas de desempenho" baseadas no folclore cult, ou até
+generalizar demais a partir de sua própria experiência. Cada bug de desempenho precisa ser
+abordado por seus próprios méritos. Mesmo se algo funcionou anteriormente, certifique-se de criar um perfil para garantir que a correção ainda seja aplicável. Seu trabalho  pode orientá-lo, mas não aplique otimizações anteriores às cegas.
+
+A melhoria do programa é um processo iterativo. Continue revisitando seu código e vendo
+que mudanças podem ser feitas. Verifique se você está progredindo em cada etapa.
+Freqüentemente, uma melhoria permitirá que outras sejam feitas. (Agora que eu não estou
+fazendo A, eu posso simplificar B fazendo C em vez disso.) Isso significa que você precisa manter-se
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
fazendo A, eu posso simplificar B fazendo C em vez disso). Isso significa que você precisa manter-se
```

File: performance-ptbr.md [hunk 5]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+olhando para a foto inteira e não fique muito obcecado com um pequeno conjunto de
+linhas
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
linhas.
```

File: performance-ptbr.md [hunk 6]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+
+Depois de escolher o algoritmo certo, a melhoria do programa é o processo de
+melhorar a implementação desse algoritmo. Na notação Grande-O, isso é
+o processo de redução das constantes associadas ao seu programa.
+
+Toda a melhoria do programa é uma das seguintes: Tornar uma coisa lenta em algo mais rápido ou fazer algo que é lento a menor quantidade de vezes.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Toda a melhoria no programa ou vai tornar mais rápido algo que era mais lento ou fazer algo que é lento uma menor quantidade de vezes.
```

File: performance-ptbr.md [hunk 7]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+As mudanças algorítmicas também se enquadram nessas categorias, mas veremos mudanças menores. Exatamente como você faz isso varia conforme as tecnologias mudam.
+
+Fazer uma coisa lenta rapidamente pode estar substituindo SHA1 ou `hash/fnv1` por uma função hash mais rápida. Fazer um processo lento menos vezes pode salvar o resultado do cálculo de hash de um arquivo grande, para que você não precise fazer isso várias vezes.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Como exemplo, fazer uma coisa lenta rapidamente pode ser substituir SHA1 ou `hash/fnv1` por uma função hash mais rápida. Fazer um processo lento menos vezes pode ser salvar o resultado do cálculo de hash de um arquivo grande para que você não precise fazer isso várias vezes.
```

File: performance-ptbr.md [hunk 8]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+>
+> É fácil ser rápido se você não precisa estar correto.
+
+A "correção" pode depender do problema. Os algoritmos heurísticos mais corretos na maioria das vezes podem ser rápidos, assim como os algoritmos que adivinham e melhoram, permitindo que você pare quando atingir um limite aceitável.
+
+
+Casos comuns de cache:
+
+Todos conhecemos o memcache, mas também existem caches em processo. O uso de um cache em processo economiza o custo da chamada de rede e o custo da serialização. Por outro lado, isso aumenta a pressão do GC, pois há mais memória para acompanhar. Você também precisa considerar estratégias de remoção, invalidação de cache e segurança de threads. 
+Um cache externo geralmente lida com o despejo, mas a invalidação do cache permanece um problema. A segurança da thread também pode ser um problema com caches externos, pois se torna um estado mutável efetivamente compartilhado entre goroutines diferentes no mesmo serviço ou até diferentes instâncias de serviço se o cache externo for compartilhado.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Um cache externo geralmente lida com o despejo, mas a invalidação do cache permanece um problema. As condições de corrida também podem ser um problema com caches externos, pois se torna um estado mutável efetivamente compartilhado entre goroutines diferentes no mesmo serviço ou até diferentes instâncias de serviço se o cache externo for compartilhado.
```

File: performance-ptbr.md [hunk 9]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+> Programas vazios dão a resposta errada em pouco tempo.
+>
+> É fácil ser rápido se você não precisa estar correto.
+
+A "correção" pode depender do problema. Os algoritmos heurísticos mais corretos na maioria das vezes podem ser rápidos, assim como os algoritmos que adivinham e melhoram, permitindo que você pare quando atingir um limite aceitável.
+
+
+Casos comuns de cache:
+
+Todos conhecemos o memcache, mas também existem caches em processo. O uso de um cache em processo economiza o custo da chamada de rede e o custo da serialização. Por outro lado, isso aumenta a pressão do GC, pois há mais memória para acompanhar. Você também precisa considerar estratégias de remoção, invalidação de cache e segurança de threads.
```
Comment (danielfireman) [inline performance-ptbr.md:512]:
```suggestion
Todos conhecemos o memcache, mas também existem caches em processo. O uso de um cache em processo economiza o custo da chamada de rede e o custo da serialização. Por outro lado, isso aumenta a pressão do GC, pois há mais memória para acompanhar. Você também precisa considerar estratégias de remoção, invalidação de cache e segurança de threads (condições de corrida). 
```

File: performance-ptbr.md [hunk 10]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+
+
+Casos comuns de cache:
+
+Todos conhecemos o memcache, mas também existem caches em processo. O uso de um cache em processo economiza o custo da chamada de rede e o custo da serialização. Por outro lado, isso aumenta a pressão do GC, pois há mais memória para acompanhar. Você também precisa considerar estratégias de remoção, invalidação de cache e segurança de threads. 
+Um cache externo geralmente lida com o despejo, mas a invalidação do cache permanece um problema. A segurança da thread também pode ser um problema com caches externos, pois se torna um estado mutável efetivamente compartilhado entre goroutines diferentes no mesmo serviço ou até diferentes instâncias de serviço se o cache externo for compartilhado.
+
+Um cache salva as informações que você acabou de gastar em tempo de computação, na esperança de poder reutilizá-las novamente em breve e economizar tempo. Um cache não precisa ser complexo. Mesmo o armazenamento de um único item - a consulta / resposta vista mais recentemente - pode ser uma grande vitória, como visto no exemplo `time.Parse ()` abaixo.
```
Comment (danielfireman) [inline performance-ptbr.md:515]:
```suggestion
Um cache salva as informações que você acabou de gastar tempo computando, na esperança de poder reutilizá-las novamente em breve e economizar tempo. Um cache não precisa ser complexo. Mesmo o armazenamento de um único item - a consulta / resposta vista mais recentemente - pode ser uma grande vitória, como visto no exemplo `time.Parse ()` abaixo.
```

File: performance-ptbr.md [hunk 11]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+
+
+A taxa de acertos esperados do cache é importante. Você deseja exportar a proporção para sua pilha de monitoramento. A alteração das taxas mostrará uma mudança no tráfego. 
+Chegou a hora de revisitar o tamanho do cache ou a política de expiração.
+
+Um cache grande pode aumentar a pressão do GC. No extremo (pouca ou nenhuma remoção, cache de todas as solicitações para uma função cara), isso pode se transformar em [memorização] (https://en.wikipedia.org/wiki/Memoization)
+
+Melhoria do programa:
+
+Melhoria do programa é a arte de melhorar iterativamente um programa em pequenas etapas.
```
Comment (danielfireman) [inline performance-ptbr.md:534]:
```suggestion
Otimizar o programa é a arte de melhorar iterativamente um programa em pequenas etapas.
```

File: performance-ptbr.md [hunk 12]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+  * Observe que implementar e otimizar cálculos numéricos é quase seu próprio campo
+    * <https://github.com/golang/go/commit/ed6c6c9c11496ed8e458f6e0731103126ce60223>
+    * <https://gist.github.com/dgryski/67e6a7ff94c3a1add30eb26ec0ad8b0f>
+    * multiplicação com adição
+    * use WolframAlpha, Maxima, sympy e similares para especializar, otimizar ou criar tabelas de pesquisa
+    * (Além disso, https://users.ece.cmu.edu/~franzf/papers/gttse07.pdf)
+    * movendo-se da matemática do ponto flutuante para a matemática inteira
+    * ou mandelbrot removendo sqrt ou lttb removendo abs, `a <b / c` =>` a * c <b`
+    * considere diferentes representações numéricas: ponto fixo, ponto flutuante, números inteiros (menores),
+    * extravagante: números inteiros com acumuladores de erro (por exemplo, linha e círculo de Bresenham), números com várias bases / sistemas de números redundantes
```
Comment (danielfireman) [inline performance-ptbr.md:559]:
```suggestion
    * sofisticado: números inteiros com acumuladores de erro (por exemplo, linha e círculo de Bresenham), números com várias bases / sistemas de números redundantes
```

File: performance-ptbr.md [hunk 13]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+  * "pague apenas pelo que você usa, não pelo que você poderia ter usado"
+    * zero apenas parte de uma matriz, em vez da coisa toda
```
Comment (danielfireman) [inline performance-ptbr.md:561]:
```suggestion
    * zero apenas parte de uma sequência, em vez da coisa toda
```

File: performance-ptbr.md [hunk 14]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+  * melhor feito em pequenas etapas, algumas declarações de cada vez
```
Comment (danielfireman) [inline performance-ptbr.md:562]:
```suggestion
  * melhor ser feito em pequenas etapas, poucas declarações de cada vez
```

File: performance-ptbr.md [hunk 15]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+  * checagens baratas antes de checagens caras:
+    * por exemplo, strcmp antes da regexp, (q.v., filtro de bloom antes da consulta)
+    "faça coisas caras menos vezes"
+  * casos comuns antes de casos raros
+    ou seja, evite testes extras que sempre falham
+  * desenrolar ainda eficaz: https://play.golang.org/p/6tnySwNxG6O
```
Comment (danielfireman) [inline performance-ptbr.md:568]:
```suggestion
  * desenrolar ainda é eficaz: https://play.golang.org/p/6tnySwNxG6O
```

File: performance-ptbr.md [hunk 16]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+    * tamanho do código. sobrecarga de teste de ramificação vs
```
Comment (danielfireman) [inline performance-ptbr.md:569]:
```suggestion
    * tamanho do código versus a sobrecarga de teste de ramificação (branch test)
```

File: performance-ptbr.md [hunk 17]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+  * usar deslocamentos em vez de atribuição de fatia pode ajudar com verificações de limites, dependências de dados e geração de código (menos para copiar no loop interno).
+  * remova verificações de limites e verificações nulas dos loops: https://go-review.googlesource.com/c/go/+/151158
+  * outros truques para o passe de prova
```
Comment (danielfireman) [inline performance-ptbr.md:572]:
```suggestion
  * outros truques para passar a prova
```

File: performance-ptbr.md [hunk 18]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+  * é aqui que pedaços de prazer para um hacker caem
```
Comment (danielfireman) [inline performance-ptbr.md:573]:
```suggestion
  * é aqui que os pedaços de prazer para um hacker caem
```

File: performance-ptbr.md [hunk 19]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+
+Muitas dicas de desempenho do folclore para ajuste dependem de compiladores de otimização insuficiente e incentivam o programador a fazer essas transformações manualmente. Os
```
Comment (danielfireman) [inline performance-ptbr.md:575]:
```suggestion
Muitas dicas folclóricas para otimização desempenho são baseadas em compiladores com otimização insuficiente e incentivam o programador a fazer essas transformações manualmente. Os
```

File: performance-ptbr.md [hunk 20]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+compiladores usam turnos em vez de multiplicar ou dividir por uma potência de dois há 15 anos - ninguém deve fazer isso manualmente. Da mesma forma, içar
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
compiladores usam deslocamentos de bits em vez de multiplicar ou dividir por uma potência de dois há 15 anos - ninguém deve fazer isso manualmente. Da mesma forma, içar
```

File: performance-ptbr.md [hunk 21]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+
+
+Digamos que estamos processando um grande arquivo de log por um único dia,
+e cada linha começa com um carimbo de hora.
+
+```
+Sun  4 Mar 2018 14:35:09 PST <...........................>
+```
+
+For each line, we're going to call `time.Parse()` to turn it into a epoch. If
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Para cada linha, nós iremos chamar `time.Parse()` para transformá-la em época. Se
```

File: performance-ptbr.md [hunk 22]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+profiling shows us `time.Parse()` is the bottleneck, we have a few options to
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
o perfilamento nos mostra que `time.Parse()`  é um gargalo, nós temos algumas opções para
```

File: performance-ptbr.md [hunk 23]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+speed things up.
```
Comment (danielfireman) [inline performance-ptbr.md:610]: melhorar a velocidade de execução.

File: performance-ptbr.md [hunk 24]
```diff
@@ -462,3 +462,549 @@ Use média geométrica para comparar grupos de benchmarks.
...
+
+The easiest is to keep a single-item cache of the previously seen time stamp
+and the associated epoch.  As long as our log file has multiple lines for a single
+second, this will be a win.  For the case of a 10 million line log file,
+this strategy reduces the number of expensive calls to `time.Parse()` from
+10,000,000 to 86400 -- one for each unique second.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
A mais fácil é manter um cache contendo uma única linha, o tempo anterior e a época associada. Enquanto o nosso arquivo de log tiver múltiplas linhas para um mesmo segundo, teremos uma vitória. Para o caso de um arquivo de log com 10 milhões de linhas, essa estratégia reduz o número de chamadas caras `time.Parse()` de 10.000.000 to 86.400 -- uma para cada segundo único.
```

**Review by danielfireman — APPROVED** [review]
There is still a lot to review, but we already have a lot of good stuff that many PT-BR speakers could deeply benefit from. Lets apply those changes, roll out the first PT-BR version and collaboratively improve it on the go.

Thanks a lot, @dialmformart !

**Pull request merged**